channeldb/revocation_log.go

```diff
@@ -0,0 +1,547 @@
+	//
+	// Deprecated: This bucket is kept for read-only in case the user
+	// choose not to migrate the old data.
+	revocationLogBucketDeprecated = []byte("revocation-log-key")
+
+	// revocationLogBucket is a sub-bucket under openChannelBucket. This
+	// sub-bucket is dedicated for storing the minimal info required to
+	// re-construct a past state in order to punish a counterparty
+	// attempting a non-cooperative channel closure.
+	revocationLogBucket = []byte("revocation-log")
+
+	// ErrLogEntryNotFound is returned when we cannot find a log entry at
+	// the height requested in the revocation log.
+	ErrLogEntryNotFound = errors.New("log entry not found")
+
+	// ErrOutputIndexTooBig is returned when the output index is greater
+	// than uint16.
+	ErrOutputIndexTooBig = errors.New("output index is over uint16")
+)
+
+// HTLCEntry specifies the minimal info needed to be stored on disk for ALL the
+// historical HTLCs, which is useful for constructing RevocationLog when a
+// breach is detected.
+// The actual size of each HTLCEntry varies based on its RHash and Amt(sat),
+// summarized as follows,
+//
+//   | RHash empty | Amt<=252 | Amt<=65,535 | Amt<=4,294,967,295 | otherwise |
+//   |:-----------:|:--------:|:-----------:|:------------------:|:---------:|
+//   |     true    |    19    |      21     |         23         |     26    |
+//   |     false   |    51    |      53     |         55         |     58    |
+//
+// So the size varies from 19 bytes to 58 bytes, where most likely to be 23 or
+// 55 bytes.
+//
+// NOTE: all the fields saved to disk use the primitive go types so they can be
+// made into tlv records without further conversion.
+type HTLCEntry struct {
+	// RHash is the payment hash of the HTLC.
+	RHash [32]byte
+
+	// RefundTimeout is the absolute timeout on the HTLC that the sender
+	// must wait before reclaiming the funds in limbo.
+	RefundTimeout uint32
+
+	// OutputIndex is the output index for this particular HTLC output
+	// within the commitment transaction.
+	//
+	// NOTE: we use uint16 instead of int32 here to save us 2 bytes, which
+	// gives us a max number of HTLCs of 65K.
+	OutputIndex uint16
+
+	// Incoming denotes whether we're the receiver or the sender of this
+	// HTLC.
+	//
+	// NOTE: this field is the memory representation of the field
+	// incomingUint.
+	Incoming bool
+
+	// Amt is the amount of satoshis this HTLC escrows.
+	//
+	// NOTE: this field is the memory representation of the field amtUint.
+	Amt btcutil.Amount
+
+	// amtTlv is the uint64 format of Amt. This field is created so we can
+	// easily make it into a tlv record and save it to disk.
+	//
+	// NOTE: we keep this field for accounting purpose only. If the disk
+	// space becomes an issue, we could delete this field to save us extra
+	// 8 bytes.
+	amtTlv uint64
+
+	// incomingTlv is the uint8 format of Incoming. This field is created
+	// so we can easily make it into a tlv record and save it to disk.
+	incomingTlv uint8
```

Comment:
Funny you brought this up as initially I had a draft that used a very different design. Basically I stored all HTLCs in one sub-bucket using the htlc indexes and serialized their indexes inside a revocation log. I did this because I mistakenly thought there might be overlaps of htlcs among different commitments so I could save more space🙈

I think the biggest issue to rely on revocation logs as the source of truth is that, the truth is deleted when the channel is marked as fully closed. So we only have access to the logs when the channel is still open/pending-close. That probably implies the app builds on top of LND needs to store that data somewhere anyway?

> Settling an invoice isn't the final point, because if LND goes offline for a few blocks (long enough to let the htlc expire) directly after, it could be that the HTLC is timed out on chain. 

> So perhaps invoices shouldn't be marked settled before the settle is locked-in (off-chain) or the htlc swept (on-chain)?

Trying to understand the issue more here. I did see we `NotifyExitHopHtlc` that settles the invoice before we `processHtlcResolution` that settles the htlc. Maybe they should be switched?

Reply:
>I think the biggest issue to rely on revocation logs as the source of truth is that, the truth is deleted when the channel is marked as fully closed. So we only have access to the logs when the channel is still open/pending-close. That probably implies the app builds on top of LND needs to store that data somewhere anyway?

Yes that's a good point.

An external app could maybe store the data, but then there still needs to be the right data coming from lnd to have something to store. Some kind of a notification stream for the actual moment of settling, but then implemented in a reliable manner so that it is impossible to miss events. Also if the app is down, what happens to the data if the channel is closed in the mean time?

It seems to me that a basic log of all htlcs settled should persist after closing the channel. I think that lnd is the only process that can reliably track this and it is critical information.

>Trying to understand the issue more here. I did see we NotifyExitHopHtlc that settles the invoice before we processHtlcResolution that settles the htlc. Maybe they should be switched?

Yes, that is indeed the case. Maybe switched, or split the invoice states into "settle requested" and "settled".